website/project/views/node.py

```diff
@@ -833,6 +844,107 @@ def _get_children(node, auth, indent=0):
+        view_timestamp = view_timestamp.get(root_id, default_timestamp)
+    return Comment.find(Q('node', 'eq', node) &
+                        Q('user', 'ne', user) &
+                        Q('date_created', 'gt', view_timestamp) &
+                        Q('date_modified', 'gt', view_timestamp) &
+                        Q('is_deleted', 'eq', False) &
+                        Q('is_hidden', 'eq', False) &
+                        Q('root_target', 'eq', root_target)).count()
+
+
+def n_unread_total(node, user, page, check=False):
+    if not node.is_contributor(user):
+        return 0
+    if page == 'node':
+        return n_unread_total_node(user, node)
+    elif page == 'files':
+        return n_unread_total_files(user, node, check=check)
+    elif page == 'wiki':
+        return n_unread_total_wiki(user, node)
+    return n_unread_total_node(user, node) + n_unread_total_wiki(user, node) + n_unread_total_files(user, node, check=check)
+
+
+def n_unread_total_node(user, node):
+    default_timestamp = datetime(1970, 1, 1, 12, 0, 0)
+    view_timestamp = user.comments_viewed_timestamp.get(node._id, dict())
+    view_timestamp = view_timestamp.get('node', None)
+    if not view_timestamp:
+        user.comments_viewed_timestamp[node._id] = dict()
+        user.comments_viewed_timestamp[node._id]['node'] = default_timestamp
+        view_timestamp = default_timestamp
+    return Comment.find(Q('node', 'eq', node) &
+                        Q('user', 'ne', user) &
+                        Q('date_created', 'gt', view_timestamp) &
+                        Q('date_modified', 'gt', view_timestamp) &
+                        Q('is_deleted', 'eq', False) &
+                        Q('is_hidden', 'eq', False) &
+                        Q('page', 'eq', 'node')).count()
+
+
+def n_unread_total_wiki(user, node):
+    from website.addons.wiki.model import NodeWikiPage
+    root_targets = NodeWikiPage.find(Q('node', 'eq', node)).get_keys()
+    n_unread = 0
+    for root_target in root_targets:
+        wiki_page = NodeWikiPage.load(root_target)
+        if hasattr(wiki_page, 'commented'):
+            root_id = wiki_page.page_name
+            n_unread += n_unread_comments(node, user, 'wiki', root_id)
+    return n_unread
+
+
+def n_unread_total_files(user, node, check=False):
+    default_timestamp = datetime(1970, 1, 1, 12, 0, 0)
+    view_timestamp = user.comments_viewed_timestamp.get(node._id, dict())
+    view_timestamp = view_timestamp.get('files', default_timestamp)
+    n_unread = 0
+    if isinstance(view_timestamp, dict):
+        for file_id in node.commented_files.keys():
+            n_unread += n_unread_comments(node, user, 'files', file_id, check)
+    else:
+        file_timestamps = dict()
+        user.comments_viewed_timestamp[node._id]['files'] = file_timestamps
+        for file_id in node.commented_files.keys():
+            file_timestamps[file_id] = default_timestamp
```

Comment:
This is modifying the database records, which is inconsistent with the other `n_unread_*` functions (the name `n_unread` implies that there won't be any side-effects/writes). Can this be refactored?